### PR TITLE
JP-3785:  Add tangent plane transform corrections in NIRSpec spectral resampling

### LINKED ISSUE #8905: NIRSpec resampled WCS placing extraction position in wrong location
_Issue [JP-3785](https://jira.stsci.edu/browse/JP-3785) was created on JIRA by [Christian Hayes](https://jira.stsci.edu/secure/ViewProfile.jspa?name=chayes):_

With the new NIRSpec spectral resampling in jwst 1.16.0, we are seeing that the extraction boxes for point sources are now missing some point sources when using use_source_posn=True (the default) in extract_1d.  This setting uses the source_ypos "known" position of the source (which hasn't changed since the previous version) and projects it onto the detector frame, so it seems the issue is related to the new WCS.  The attached plots show the s2d spectrum (upper panel) with the extraction box highlighted by the red lines (the gray dashed lines are not relevant but show the center of the s2d data array which would be used with use_source_posn=False), and the x1d spectrum (lower panel) and its error spectrum (black and red lines respectively).  Previously this source (PID 1747 obs 16 source 731) was well centered in the extraction box, but now it is not.

### PR DESCRIPTION
**Pull request by hayescr**
Resolves [JP-3785](https://jira.stsci.edu/browse/JP-3785)

Closes #8905

This PR fixes an issue with the NIRSpec spectral resampled WCS transforms from resampled pixels to a sky tangent projection.  The transforms account for the additional extent of nodded slits along the tangent plane but the code currently only adds this correction for the sky direction that the slit is better aligned with.  This PR adds the necessary corrections for both tangent plane dimensions.

## Tasks
- [x] **request a review from someone specific**, to avoid making the maintainers review every PR
- [x] add a build milestone, i.e. `Build 11.3` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [x] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [x] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types) 
  - [ ] update or add relevant tests
  - [ ] update relevant docstrings and / or `docs/` page
  - [x] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [x] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
- ``changes/<PR#>.docs.rst``
- ``changes/<PR#>.stpipe.rst``
- ``changes/<PR#>.datamodels.rst``
- ``changes/<PR#>.scripts.rst``
- ``changes/<PR#>.fits_generator.rst``
- ``changes/<PR#>.set_telescope_pointing.rst``
- ``changes/<PR#>.pipeline.rst``

## stage 1
- ``changes/<PR#>.group_scale.rst``
- ``changes/<PR#>.dq_init.rst``
- ``changes/<PR#>.emicorr.rst``
- ``changes/<PR#>.saturation.rst``
- ``changes/<PR#>.ipc.rst``
- ``changes/<PR#>.firstframe.rst``
- ``changes/<PR#>.lastframe.rst``
- ``changes/<PR#>.reset.rst``
- ``changes/<PR#>.superbias.rst``
- ``changes/<PR#>.refpix.rst``
- ``changes/<PR#>.linearity.rst``
- ``changes/<PR#>.rscd.rst``
- ``changes/<PR#>.persistence.rst``
- ``changes/<PR#>.dark_current.rst``
- ``changes/<PR#>.charge_migration.rst``
- ``changes/<PR#>.jump.rst``
- ``changes/<PR#>.clean_flicker_noise.rst``
- ``changes/<PR#>.ramp_fitting.rst``
- ``changes/<PR#>.gain_scale.rst``

## stage 2
- ``changes/<PR#>.assign_wcs.rst``
- ``changes/<PR#>.badpix_selfcal.rst``
- ``changes/<PR#>.msaflagopen.rst``
- ``changes/<PR#>.nsclean.rst``
- ``changes/<PR#>.imprint.rst``
- ``changes/<PR#>.background.rst``
- ``changes/<PR#>.extract_2d.rst``
- ``changes/<PR#>.master_background.rst``
- ``changes/<PR#>.wavecorr.rst``
- ``changes/<PR#>.srctype.rst``
- ``changes/<PR#>.straylight.rst``
- ``changes/<PR#>.wfss_contam.rst``
- ``changes/<PR#>.flatfield.rst``
- ``changes/<PR#>.fringe.rst``
- ``changes/<PR#>.pathloss.rst``
- ``changes/<PR#>.barshadow.rst``
- ``changes/<PR#>.photom.rst``
- ``changes/<PR#>.pixel_replace.rst``
- ``changes/<PR#>.resample_spec.rst``
- ``changes/<PR#>.residual_fringe.rst``
- ``changes/<PR#>.cube_build.rst``
- ``changes/<PR#>.extract_1d.rst``
- ``changes/<PR#>.resample.rst``

## stage 3
- ``changes/<PR#>.assign_mtwcs.rst``
- ``changes/<PR#>.mrs_imatch.rst``
- ``changes/<PR#>.tweakreg.rst``
- ``changes/<PR#>.skymatch.rst``
- ``changes/<PR#>.exp_to_source.rst``
- ``changes/<PR#>.outlier_detection.rst``
- ``changes/<PR#>.tso_photometry.rst``
- ``changes/<PR#>.stack_refs.rst``
- ``changes/<PR#>.align_refs.rst``
- ``changes/<PR#>.klip.rst``
- ``changes/<PR#>.spectral_leak.rst``
- ``changes/<PR#>.source_catalog.rst``
- ``changes/<PR#>.combine_1d.rst``
- ``changes/<PR#>.ami.rst``

## other
- ``changes/<PR#>.wfs_combine.rst``
- ``changes/<PR#>.white_light.rst``
- ``changes/<PR#>.cube_skymatch.rst``
- ``changes/<PR#>.engdb_tools.rst``
- ``changes/<PR#>.guider_cds.rst``
</details>

### CODE REVIEW

**Review by melanieclarke — COMMENTED** [review]
One small suggestion for refactoring a helper function to avoid repeated calculations.  Also, please add a change note - see instructions in the PR template, or let me know if you have difficulty with the new format.

File: jwst/resample/resample_spec.py [hunk 1]
```diff
@@ -335,32 +335,45 @@ def build_nirspec_output_wcs(self, input_models, refmodel=None):
...
-        if swap_xy:
-            pix_to_tan_slope = np.abs(pix_to_ytan.slope)
-            slope_sign = np.sign(pix_to_ytan.slope)
-        else:
-            pix_to_tan_slope = np.abs(pix_to_xtan.slope)
-            slope_sign = np.sign(pix_to_xtan.slope)
+        min_tan_y, max_tan_y = self._max_spatial_extent(
+            all_wcs, undist2sky.inverse, True)
+        min_tan_x, max_tan_x = self._max_spatial_extent(
+            all_wcs, undist2sky.inverse, False)
```
[the remark this answers is not in the document]
Comment (hayescr) [reply jwst/resample/resample_spec.py]: Thanks for the suggestion, I've made some changes, which I'm not sure are necessarily better, but it at least reduces some of the repeated computations.

**Comment by melanieclarke** [comment]
Regression tests started here:
https://github.com/spacetelescope/RegressionTests/actions/runs/11469829363

**Comment by melanieclarke** [comment]
Regression test results look as expected for this change.  There are changes to most NIRSpec MOS and FS products downstream of resampling.  

The s2d changes are all changes to the WCS only, which shows up in diffs in the S_REGIONs.  The x1d products show mostly fractional pixel changes to the extraction location (EXTRYSTR, EXTRYSTP) and corresponding minor changes to the extracted flux.

The MOS/FS test data set is the only one in the regtest suite that shows a significant change to the extraction location: test_nirspec_mos_fs_spec3[s000001354-x1d] shows a shift of about 1.5 pixels in extraction location.  Looking at the corresponding s2d file, the new aperture location looks like an improvement over the old one: it centers the source better.

jw02674-o004_s000001354_nirspec_f290lp-g395m_x1d

**Review by melanieclarke — APPROVED** [review]
LGTM!  Thanks again for catching this and providing the fix.

**Pull request merged**